tests/util.go

```diff
@@ -10,108 +10,123 @@ import (
+func (b *CmdBuilder) FilePath(filePath string) *CmdBuilder {
+	if b.cmd == "virtctl" {
+		ktests.PanicOnError(fmt.Errorf("virtctl doesn't support -f currently"))
+		return nil
+	}
+	b.filePath = filePath
+	b.cmdArgs = append(b.cmdArgs, "-f", filePath)
+	return b
 }
 
-func exec(r Result) {
-	var err error
-	if r.verb == "" {
-		Expect(fmt.Errorf("verb can not be empty"))
-	}
-	cmd := []string{r.verb}
-	if r.filePath == "" {
-		if r.resourceType == "" {
-			Expect(fmt.Errorf("resourceType can not be empty"))
-		}
-		cmd = append(cmd, r.resourceType)
-	}
-	if r.resourceName != "" {
-		cmd = append(cmd, r.resourceName)
-	}
-	if r.filePath != "" {
-		cmd = append(cmd, "-f", r.filePath)
-	}
-	if r.resourceLabel != "" {
-		cmd = append(cmd, "-l", r.resourceLabel)
-	}
-	if r.query != "" {
-		cmd = append(cmd, r.query)
-	}
-	if r.nameSpace != "" {
-		cmd = append(cmd, "-n", r.nameSpace)
-	}
+func (b *CmdBuilder) ResourceType(resourceType string) *CmdBuilder {
+	b.resourceType = resourceType
+	b.cmdArgs = append(b.cmdArgs, resourceType)
+	return b
+}
+
+func (b *CmdBuilder) ResourceName(resourceName string) *CmdBuilder {
+	b.resourceName = resourceName
+	b.cmdArgs = append(b.cmdArgs, b.resourceName)
+	return b
+}
 
-	if r.expectOut != "" {
-		Eventually(func() bool {
-			r.actualOut, err = ktests.RunOcCommand(cmd...)
-			Expect(err).ToNot(HaveOccurred())
-			return strings.Contains(r.actualOut, r.expectOut)
-		}, time.Duration(2)*time.Minute).Should(BeTrue(), fmt.Sprintf("Timed out waiting for %s to appear", r.resourceType))
-	} else {
-		r.actualOut, err = ktests.RunOcCommand(cmd...)
+func (b *CmdBuilder) ResourceLabel(resourceLabel string) *CmdBuilder {
+	b.resourceLabel = resourceLabel
+	b.cmdArgs = append(b.cmdArgs, "-l", resourceLabel)
+	return b
+}
+
+func (b *CmdBuilder) WaitUntil(expectOut string) {
+	Eventually(func() bool {
+		spew.Dump(b.cmdArgs)
+		out, err := ktests.RunCommand(b.cmdArgs[0], b.cmdArgs[1:]...)
 		Expect(err).ToNot(HaveOccurred())
-	}
+		return strings.Contains(out, expectOut)
+	}, time.Duration(2)*time.Minute).Should(BeTrue(), fmt.Sprintf("Timed out waiting for %s to appear", b.resourceType))
 }
 
-func writeJson(jsonFile string, json string) (string, error) {
-	err := ioutil.WriteFile(jsonFile, []byte(json), 0644)
+func (b *CmdBuilder) WriteJson(jsonFile string) {
+	var err error
+	b.actualOut, err = ktests.RunCommand(b.cmdArgs[0], b.cmdArgs[1:]...)
+	Expect(err).ToNot(HaveOccurred())
+	err = ioutil.WriteFile(jsonFile, []byte(b.actualOut), 0644)
 	if err != nil {
-		return "", fmt.Errorf("failed to write the json file %s", jsonFile)
+		ktests.PanicOnError(fmt.Errorf("failed to write the json file %s", jsonFile))
 	}
-	return jsonFile, nil
 }
+
+func (b *CmdBuilder) Run() string {
+	var err error
+	b.actualOut, err = ktests.RunCommand(b.cmdArgs[0], b.cmdArgs[1:]...)
+	Expect(err).ToNot(HaveOccurred())
+	return b.actualOut
```

Comment:
I would create CmdResult struct and set there all results (stdout, stderr, rc), define some action over that structure, like expectOutput(XY), exitsWith(RC) ... so caller of this function can control what is correct / expected result of executed command.